test/mpi/threads/pt2pt/mt_probe_impl.c

```diff
@@ -295,7 +295,14 @@ MTEST_THREAD_RETURN_TYPE run_test(void *arg)
 #else
     MTestPrintfMsg(2, "Testing probe_normal\n");
     p->result = test_probe_normal(p->id, p->iter, p->count, p->buff, p->comm, p->tag, p->verify);
-    MPI_Barrier(p->comm);
+
+    /* Barrier across threads. */
+    MTest_thread_barrier(NTHREADS);
```

Comment:
Testing on FreeBSD, it does fix the occasional hang. Further testing shows that only this `MTest_thread_barrier` is crucial -- i.e. just add a thread_barrier before the `MPI_Barrier` won't hang.

So it indicate we have some issue of `MPI_Barrier` mix up some message from `test_probe_normal`, which indicate an issue of context_id mask mess up.

Reply:
I believe my fix is correct. I do not think it is safe to call `MPI_Barrier(MPI_COMM_WORLD);` from multiple threads simultaneously like this:
```
#omp parallel for num_threads(16)
for(int i = 0; i < 16; i++)
  MPI_Barrier(MPI_COMM_WORLD);
```

Reply:
The original code was not doing that.  The original code was doing a barrier on a different communicator for each thread, which is correct.  I don't think your patch is correct.

Reply:
There are 3 batch of tests for this series: 
* same comm_world, same tag
* same comm_world, different tag
* different comm, same tag

So for the first two batch, I believe it is call MPI_Barrier on the same communicator. Nevertheless, I believe the test failure is from mixing pt2pt messages with the barrier messages, somehow, on freebsd -- it needs investigate.

Reply:
I checked the history of this file, but I could not find the original code you mention.
https://github.com/pmodels/mpich/blob/master/test/mpi/threads/pt2pt/mt_probe_impl.c

Anyway, I believe you are referring to the following:
https://github.com/pmodels/mpich/blob/master/test/mpi/threads/pt2pt/mt_pt2pt_common.c#L165

`mt_pt2pt_common.c` has other paths, however. If we need a different communicator for each thread, I need to change the following part:
https://github.com/pmodels/mpich/blob/master/test/mpi/threads/pt2pt/mt_pt2pt_common.c#L114
https://github.com/pmodels/mpich/blob/master/test/mpi/threads/pt2pt/mt_pt2pt_common.c#L141

I can come up with a few solutions:

0. My current fix (which looks reasonable to me).
1. Skip this `run_test()` when `p_comm` points to `MPI_COMM_WORLD`, 
2. Create and free per-thread barrier in `run_test()` apart from `p_comm`.

Reply:
The bug is that `p_comm` is set to `MPI_COMM_WORLD` for all communicators.  That needs to be fixed, instead of adding thread barriers.

Reply:
@pavanbalaji Can I remove these two paths, then?
https://github.com/pmodels/mpich/blob/master/test/mpi/threads/pt2pt/mt_pt2pt_common.c#L114
https://github.com/pmodels/mpich/blob/master/test/mpi/threads/pt2pt/mt_pt2pt_common.c#L141

Reply:
What's the reason for the `MPI_Barrier` in the first place?  Can that be removed?  I think the test is trying to make sure pt2pt communication on the same communicator works fine, so we shouldn't remove that.

Reply:
I think the `MPI_Barrier` is just to synchronize between each batch of tests. 

I agree that we shouldn't remove those two intended tests -- same comm same tag and same comm per-thread tag.

Personally I think @shintaro-iwasaki 's fix is fine. It is semantically clean. However, what was causing the issue was not the parallel `MPI_Barrier` though. It is a bit non-deterministic, just like the first batch of "same comm same tag" tests, but technically should work fine. As far as my testing indicate, the issue on FreeBSD is due to some of the message from previous pt2pt leaked into barrier messages -- which should not happen unless the context_id+context_offset+tag is not working perfectly on that platform. Infact, simply having a thread_barrier before the `MPI_Barrier` makes the test failure go away. But that is just a patch. We still need understand and fix the real issue.

Reply:
No, it's technically an incorrect program and should hang/fail once in a while (which is what seems to be happening).  After looking through the code, I agree that @shintaro-iwasaki's fix is probably fine.  My original objection was incorrect.